### ISSUE TITLE
Support Schnorr signatures

### ISSUE THREAD

**Issue by schildbach** [bcgit/bc-java#1114]
(I was pointed here by the dev-crypto list, the thread is mysteriously missing from the mailing list archive.)

My feature request is to support Schnorr signatures, specifically variant that Bitcoin uses: [BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki)

David made the suggestion to add them to the signers package in the low level API and then add a suitable named equivalent to the JCE provider.

**Comment by msgilligan** [comment]
The related **bitcoinj** issue is here: https://github.com/bitcoinj/bitcoinj/issues/1949

There is some public-domain Java code in a repo hosted by Samurai, here: ~~https://code.samourai.io/samouraidev/BIP340_Schnorr~~

**Update:** See the [References section](https://github.com/bitcoinj/secp256k1-jdk#references) of the secp256k1-jdk README for a list of Schnorr Signature implementations (including a GitHub repo with the Samurai implementation.)

I'm proposing that the **bitcoinj** project start an implementation that we would hopefully, eventually submit to Bouncy Castle. Is anyone interesting in providing some guidance or helping with this effort?

**Comment by msgilligan** [comment]
> I'm proposing that the **bitcoinj** project start an implementation

We have created the [secp256k1-jdk](https://github.com/bitcoinj/secp256k1-jdk) library which creates an secp256k1-specific API for cryptographic functions including Schnorr sign/verify. There is currently no JCA provider but that could be added in the future.

There are two implementations of the API:

* A pure Java implementation using Bouncy Castle
* A native implementation using [libsecp256k1](https://github.com/bitcoin-core/secp256k1)

The Bouncy Castle implementation of Schnorr sign/verify uses the Bouncy Castle secp256k1 classes for some math, but is currently using BigInteger math for operations that were either not supported by Bouncy Castle or were not `public`. We plan to investigate further and improve the implementation. It _might_ make sense to make a PR to Bouncy Castle to add some math functionality that is helpful for Schnorr signatures and might be useful for other applications.

**Comment by msgilligan** [comment]
We have published pre-release: [v0.2.1](https://github.com/bitcoinj/secp256k1-jdk/blob/master/CHANGELOG.adoc#v021)

Binaries are in a GitLab Maven repo -- see the README for how to access them.

With Java 25 and JBang, you can run the following Java script:

```
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25+
//REPOS mavencentral, secpjdk=https://gitlab.com/api/v4/projects/55956336/packages/maven
//DEPS org.bitcoinj.secp:secp-bouncy:0.2.1

import org.bitcoinj.secp.*;

String msg = "Hello, world!";
String tag = "my_fancy_protocol";

void main() {
    IO.println("secp256k1-jdk Schnorr Sample");
    try (Secp256k1 secp = Secp256k1.getById(Secp256k1.ProviderId.BOUNCY_CASTLE)) {
        var messageHash = secp.taggedSha256(tag, msg);
        var pair = secp.ecKeyPairCreate();
        var sig = secp.schnorrSigSign32(messageHash, pair);
        var isValid = secp.schnorrSigVerify(sig, messageHash, pair.publicKey()).get();
        IO.println("Valid Schnorr signature = " + isValid);
    }
}
```

**Comment by msgilligan** [comment]
> The Bouncy Castle implementation of Schnorr sign/verify uses the Bouncy Castle secp256k1 classes for some math, but is currently using BigInteger math for operations that were either not supported by Bouncy Castle or were not `public`. We plan to investigate further and improve the implementation. It _might_ make sense to make a PR to Bouncy Castle to add some math functionality that is helpful for Schnorr signatures and might be useful for other applications.

@peterdettman any thoughts on this?

**Comment by MutantRabbit767** [comment]
I started working on a Nostr chat app about 3 days ago. This is going to seriously save me since I've been absolutely STRUGGLING to get schnorr signatures to work. Thank you everyone for your hard work.

I'll need to figure out how to get this working in my Kotlin application.

**Comment by msgilligan** [comment]
@MutantRabbit767 asks:

> I'll need to figure out how to get this working in my Kotlin application.

Take a look at:

https://github.com/bitcoinj/secp256k1-jdk/blob/master/secp-examples-kotlin/src/main/java/org/bitcoinj/secp/kotlin/examples/Schnorr.kt

**Comment by msgilligan** [comment]
Looks like some work on Schnorr / BIP340 is happening:
https://github.com/bcgit/bc-java/blob/main/core/src/main/java/org/bouncycastle/crypto/signers/BIP340Signer.java

**Comment by ligefeiBouncycastle** [comment]
Thanks for the nudge, and apologies for how long this one sat open. BIP-340 Schnorr signatures over secp256k1 are now implemented as a lightweight signer (org.bouncycastle.crypto.signers.BIP340Signer), verified against the official bitcoin/bips test vectors. It's available in the latest beta — https://www.bouncycastle.org/betas — and will ship in the next release.

Closing this out, but please give the beta a try and open a new issue or reopen this issue if anything looks off.

**Comment by msgilligan** [comment]
> Thanks for the nudge, and apologies for how long this one sat open. BIP-340 Schnorr signatures over secp256k1 are now implemented as a lightweight signer (org.bouncycastle.crypto.signers.BIP340Signer), verified against the official bitcoin/bips test vectors. It's available in the latest beta — https://www.bouncycastle.org/betas — and will ship in the next release.
> 

@ligefeiBouncycastle @peterdettman Unfortunately, we've been working in parallel and we just released [secp256k1-jdk](https://github.com/bitcoinj/secp256k1-jdk) (an API with native/libsecp256k1 and a [Bouncy Castle implementation](https://github.com/bitcoinj/secp256k1-jdk/tree/master#bouncy-castle-implementation)) with a Schnorr signature using lower-level BC primitives. [Release 0.2.1](https://github.com/bitcoinj/secp256k1-jdk/blob/master/CHANGELOG.adoc#v021) (on May 7, 2026) was the first to have Bouncy Castle Schnorr sign/verify.

We were talking about trying to submit changes upstream. We are also working on MuSig2 and plan a Bouncy Castle implementation. Is there a way we can work more closely together in the future?

> Closing this out, but please give the beta a try and open a new issue or reopen this issue if anything looks off.

We will definitely compare to our implementation and hopefully replace/update it.

**Comment by ligefeiBouncycastle** [comment]
@msgilligan Thank you for your trust. On working more closely together, I'd be glad to take that forward — please feel free to email me directly at gefei.li@keyfactor.com and we can talk it through. Thanks for reaching out, and for offering to contribute it back.
